pywebdriver/plugins/scale_driver.py

```diff
@@ -0,0 +1,144 @@
+# SPDX-FileCopyrightText: 2022 Coop IT Easy SCRLfs
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+"""Class API for implementing a weighing scale."""
+
+import logging
+import time
+from abc import ABC, abstractmethod
+from contextlib import ExitStack
+from importlib import import_module
+from threading import RLock, Thread
+
+from flask import jsonify
+
+from pywebdriver import app, config, drivers
+
+from .base_driver import AbstractDriver
+
+_logger = logging.getLogger(__name__)
+
+driver = None
+
+
+class AbstractScaleDriver(Thread, AbstractDriver, ABC):
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(*args, daemon=True, **kwargs)
+
+        self.active = False
+        self.config = config
+        self.connection = None
+        self.data = {}
+        self.data_lock = RLock()
+        self.vendor_product = None
+
+    @property
+    @abstractmethod
+    def weight(self):
+        """Return the last reported weight of the scale."""
+        # Read this from self.data. Use self.lock.
+
+    @property
+    @abstractmethod
+    def scale_status(self):
+        """Return the last reported status of the scale."""
+        # Read this from self.data. Use self.lock.
+
+    @property
+    def poll_interval(self):
+        """Read the poll interval from the config."""
+        return self.config.getfloat("scale_driver", "poll_interval", fallback=0.2)
```

Comment:
this would be called every 0.2 second. i assume it is not that inefficient (but parsing strings apparently anyway), but storing the float value directly in self would anyway be much faster and cleaner.

also, i suppose that it is a property to allow sub-classes to override it, but does it make sense?

the 0.2 value comes from [the 8217 mettler toledo protocol](https://www.manualslib.com/manual/861274/Mettler-Toledo-Viva.html?page=51#manual), whose documentation says:
> There must be at least a 200-ms delay between commands to allow for processing data response
time at the scale.

i think a better option would be to have the concrete protocol classes define the default value, while allowing it to be overridden in the config file (because with ariva-s scales for example, it works with lower values).

Reply:
> also, i suppose that it is a property to allow sub-classes to override it, but does it make sense?

The property is just for syntactic sugar. i.e. `driver.poll_interval` instead of `driver.get_poll_interval()`.

Otherwise I don't know if we expect the config to change after launch. If we expect that it doesn't change change, then yes, we should probably just store it directly on self.

> i think a better option would be to have the concrete protocol classes define the default value, while allowing it to be overridden in the config file (because with ariva-s scales for example, it works with lower values).

This makes sense to me.

Reply:
i think that we can assume that the config will not change after launch, but better check in the code and with the project maintainers.

i think the poll interval should be handled in a similar way as the other parameters like the port and the baudrate, as they are more related to a specific protocol implementation than a general scale driver.

Reply:
Did this change in [717cf5c](https://github.com/akretion/pywebdriver/pull/102/commits/717cf5c6769a54eb39a9b5c9ce83fe1b712101f2)